### ISSUE TITLE
About the inference speed

### ISSUE THREAD

**Issue by yuanze1024** [crockwell/Cap3D#2]
Hi, thank you for your excellent work. It is very innovative. 
However, when I reproduce it on my single A100, I found that the inference time are imbalanced between the first BLIP-2 stage(which is to get the object name), and the second BLIP-2 stage(which is to caption coarse structure and geometry info of this object). To be specific, inference on a small dataset of 108 images costs 6.1642115116119385 and 145.6894016265869(about 1.3sec per image) sec seperately, which won't match your paper(page 15th):
<img width="1086" alt="image" src="https://github.com/crockwell/Cap3D/assets/50018861/ecd1b924-97a7-41fd-a0d3-5e733a301cf9">
**So, my questions are:**
1. How about your caption inference speed? I didn't see it in the paper.
2. Is it normal to cost 1.3sec to generate a caption on a A100? What's more, I ran it on a batch size of 5, instead of just one!!!

I've also submit an issue #https://github.com/salesforce/LAVIS/issues/418 to try to find out the reason why the second BLIP-2 is that slow and it would be appreciated if you can help me find out it. And my code snippets are in #https://github.com/salesforce/LAVIS/issues/418 if you need it to debug.

-----
BTW, in your code, it will load the pretrained ckpt 8 times. I don't know is that a miss or you just mean it.
https://github.com/crockwell/Cap3D/blob/da42c36c5a7ba50a2ebfe7367633f45955e76fb0/captioning_pipeline/caption_blip2.py#L37

**Comment by crockwell** [comment]
Hi Yuan Ze,

Thanks for your thoughtful questions! I'll respond below.
- For the one-stage captioning case, we also saw inference speed of 1.3sec per caption, albeit on an A40; resulting in 2700 iterations per hour with batch size 1 (Appendix A). Glad to hear you could fit a bigger batch size on your A100. I found running a bigger batch size on an A40 did not meaningfully speed up inference, so for simplicity I used batch size of 1.
 - For the two-stage captioning case, you have made an interesting finding! Our captioning indeed speed roughly doubles, as you notice in the paper. However, the math is slightly more complicated than doubling the one-stage speed. As you notice, there is an imbalance between the first and second part of captioning. I calculated this locally, and found about 0.4sec for part 1 and about 2.7sec for part 2. These numbers don't quite match yours, but still are much longer in stage 2 than stage 1. The reason for this is (1) stage 2 produces 5 samples, while stage 1 only produces 1 sample; and (2) it seems BLIP2 QA is slower than BLIP2 captioning. So, you would expect stage 2 to be about 5x slower than stage 1, which roughly matches my local calculations. And, part 2 alone takes longer than the 5 sample case in one-stage captioning. I'll clarify this breakdown and precise sum in our next arXiv version.
 - Re: loading multiple times. This is a good point. In reality, we wrote this to be simple in representing what was needed to replicate our process, and could probably be streamlined. If you want to make a pull request feel free! Otherwise I'll try to update it in some time.

Regards,
Chris

**Comment by yuanze1024** [comment]
Thank you for your quick replying. I'm a little busy these days, sorry for not replying in time.
The following texts are translated from Chinese.

-----
I used an A800 with 80GB of VRAM (which can be considered to have similar performance to the A100) to test the impact of different batch sizes (1, 2, 5) on the speed. In the first BLIP-2 stage, the time decreases almost linearly as the batch size increases, and the edge effect appears when the batch size is 5 (I did not test higher batch sizes because the VRAM usage was already around 60GB at batch size 5). In contrast, the time in the second stage decreases slightly as the batch size increases. I attribute this phenomenon to the first stage not reaching the GPU's bottleneck in terms of FLOPs, while the second stage almost exploits the GPU's performance (not rigorously tested). 

As supporting evidence, I found on Google that the FLOPS of the A40 is approximately half that of the A100. In your BLIP-2, the second stage took about 2.7s, while mine took 1.3s. If I understand correctly, this aligns well with the actual situation.

As for the code, I just tested it in my single card, so just simply put the `load_model_and_preprocess()` out of the `main()`. 

-----
Again, thanks for your work and replying!
